### ISSUE TITLE
Sway stalls during start in `epoll_wait`

### ISSUE THREAD

**Issue by tokyovigilante** [swaywm/sway#7457]
### Please fill out the following:
`Sway version 1.9-dev-26c33a35 (Feb 17 2023, branch 'master')`
`wlroots version 0.17.0-dev`
Fedora rawhide

- **Debug Log:**
[sway.log](https://github.com/swaywm/sway/files/10772634/sway.log)

- **Configuration File:**
Occurs with default.

- **Stack Trace:**
```
0x00007f825fb8ae72 in epoll_wait (epfd=3, events=events@entry=0x7ffe541c4060, maxevents=maxevents@entry=32, timeout=timeout@entry=-1) at ../sysdeps/unix/sysv/linux/epoll_wait.c:30
30	  return SYSCALL_CANCEL (epoll_wait, epfd, events, maxevents, timeout);
Missing separate debuginfos, use: dnf debuginfo-install llvm-libs-15.0.7-2.fc38.x86_64
#0  0x00007f825fb8ae72 in epoll_wait (epfd=3, events=events@entry=0x7ffe541c4060, maxevents=maxevents@entry=32, timeout=timeout@entry=-1) at ../sysdeps/unix/sysv/linux/epoll_wait.c:30
        sc_ret = -4
        sc_cancel_oldtype = 0
#1  0x00007f825fcfb7bb in wl_event_loop_dispatch (loop=0xcb3f10, timeout=timeout@entry=-1) at ../src/event-loop.c:1004

                      ep = {{events = 1, data = {ptr = 0x1a1ed00, fd = 27389184, u32 = 27389184, u64 = 27389184}}, {events = 1, data = {ptr = 0xcb4720, fd = 13322016, u32 = 13322016, u64 = 13322016}}, {events = 1605356817, data = {ptr = 0x200007f82, fd = 32642, u32 = 32642, u64 = 8589967234}}, {events = 0, data = {ptr = 0x800000000000000e, fd = 14, u32 = 14, u64 = 9223372036854775822}}, {events = 0, data = {ptr = 0x0, fd = 0, u32 = 0, u64 = 0}}, {events = 0, data = {ptr = 0x0, fd = 0, u32 = 0, u64 = 0}}, {events = 0, data = {ptr = 0x0, fd = 0, u32 = 0, u64 = 0}}, {events = 0, data = {ptr = 0x0, fd = 0, u32 = 0, u64 = 0}}, {events = 1411137792, data = {ptr = 0x5faf7fd200007ffe, fd = 32766, u32 = 32766, u64 = 6894870094447411198}}, {events = 32642, data = {ptr = 0x4, fd = 4, u32 = 4, u64 = 4}}, {events = 1411138520, data = {ptr = 0x400007ffe, fd = 32766, u32 = 32766, u64 = 17179901950}}, {events = 0, data = {ptr = 0x7ffe541c43d8, fd = 1411138520, u32 = 1411138520, u64 = 140730309559256}}, {events = 4, data = {ptr = 0x0, fd = 0, u32 = 0, u64 = 0}}, {events = 0, data = {ptr = 0x7ffe541c4150, fd = 1411137872, u32 = 1411137872, u64 = 140730309558608}}, {events = 4677276, data = {ptr = 0x6048400000000000, fd = 0, u32 = 0, u64 = 6937865594708426752}}, {events = 32642, data = {ptr = 0x7ffe541c4178, fd = 1411137912, u32 = 1411137912, u64 = 140730309558648}}, {events = 4716208, data = {ptr = 0x5fcf9d4500000000, fd = 0, u32 = 0, u64 = 6903909673460563968}}, {events = 2, data = {ptr = 0x0, fd = 0, u32 = 0, u64 = 0}}, {events = 576528333, data = {ptr = 0x5fcf6db000000000, fd = 0, u32 = 0, u64 = 6903857356463931392}}, {events = 32642, data = {ptr = 0x22d7062cd, fd = 762340045, u32 = 762340045, u64 = 9352274637}}, {events = 1411138112, data = {ptr = 0x475fa700007ffe, fd = 32766, u32 = 32766, u64 = 20089894210666494}}, {events = 0, data = {ptr = 0x47f6b0, fd = 4716208, u32 = 4716208, u64 = 4716208}}, {events = 855813376, data = {ptr = 0x400000002, fd = 2, u32 = 2, u64 = 17179869186}}, {events = 0, data = {ptr = 0x3000000028, fd = 40, u32 = 40, u64 = 206158430248}}, {events = 1411138128, data = {ptr = 0x541c419000007ffe, fd = 32766, u32 = 32766, u64 = 6060791285264973822}}, {events = 32766, data = {ptr = 0x7f8260484000 <_rtld_local>, fd = 1615347712, u32 = 1615347712, u64 = 140197937823744}}, {events = 4836552, data = {ptr = 0x47f4d800000000, fd = 0, u32 = 0, u64 = 20253931896569856}}, {events = 0, data = {ptr = 0x14c, fd = 332, u32 = 332, u64 = 332}}, {events = 22614384, data = {ptr = 0x2d7062cd00000000, fd = 0, u32 = 0, u64 = 3274225561706168320}}, {events = 0, data = {ptr = 0x15b4440, fd = 22758464, u32 = 22758464, u64 = 22758464}}, {events = 22757888, data = {ptr = 0x541c41f000000000, fd = 0, u32 = 0, u64 = 6060791697581801472}}, {events = 32766, data = {ptr = 0x7f8260397646 <wlr_backend_start+46>, fd = 1614378566, u32 = 1614378566, u64 = 140197936854598}}}
        source = <optimized out>
        i = <optimized out>
        count = <optimized out>
        has_timers = false
#2  0x00007f825fcfc025 in wl_display_run (display=0xcb3e20) at ../src/wayland-server.c:1431
#3  0x00000000004201c2 in server_run (server=0x49eae0 <server>) at ../sway/server.c:334
#4  0x000000000041f41d in main (argc=4, argv=0x7ffe541c43d8) at ../sway/main.c:415
        verbose = false
        debug = true
        validate = false
        allow_unsupported_gpu = false
        config_path = 0xcb3360 "/dev/null"
        c = -1
Detaching from program: /usr/local/bin/sway, process 3935
```

- **Description:**
Up to date wlroots, sway and Fedora rawhide. After reboot sway fails to start, stuck at tty1. Can switch tty and gather above information using gdb attached to process, seems to be stalled in `epoll_wait`. No obvious driver issue in `dmesg`, using Intel drivers on a Thinkpad X1 Yoga (Cannon Lake, UHD graphics).

**Comment by bl4ckb0ne** [comment]
Start sway with `-d > ~/sway.log 2>&1` to redirect all output into a file and read it next reboot.

**Comment by tokyovigilante** [comment]
Thanks for reviewing, the above log is a debug log but have attached a further log with `-d > ~/sway.log 2>&1` and capturing the reboot from another tty.
[sway.log](https://github.com/swaywm/sway/files/10772878/sway.log)

**Comment by bl4ckb0ne** [comment]
Seems to be an issue with wlroots and EGL
> 00:00:00.004 [DEBUG] [wlr] [../render/egl.c:450] EGL_EXT_device_drm not supported
> 00:00:00.027 [INFO] [wlr] [../render/egl.c:358] Supported EGL device extensions: EGL_EXT_device_drm EGL_EXT_device_drm_render_node

**Comment by emersion** [comment]
This is a regression: https://gitlab.freedesktop.org/wlroots/wlroots/-/merge_requests/4006#note_1780235

**Comment by tokyovigilante** [comment]
Have updated wlroots to https://gitlab.freedesktop.org/wlroots/wlroots/-/commit/782b5e65655d3b8d1ecdd31c57dea342a6a0c0e0 which has resolved the EGL extension detection issue but see the same behavior. New debug log attached.

Is this relevant? Seems to affect all my startup commands in config.
```
00:00:01.045 [INFO] [wlr] [../subprojects/wlroots/backend/headless/backend.c:16] Starting headless backend
00:00:01.045 [DEBUG] [../sway/config.c:682] Running deferred commands
00:00:01.045 [INFO] [../sway/commands.c:261] Handling command 'exec systemctl --user import-environment'
00:00:01.045 [DEBUG] [../sway/commands/exec_always.c:58] Executing systemctl --user import-environment
00:00:01.045 [DEBUG] [../sway/desktop/launcher.c:235] Failed to create launch context. No workspace.
```
[sway.log](https://github.com/swaywm/sway/files/10775357/sway.log)

**Comment by rpigott** [comment]
No, the launch context is only for workspace matching. It is ok to fail, since workspace matching is an optional enhancement with a simple fallback.

**Comment by tokyovigilante** [comment]
Ok thanks for clarifying. Have rolled back to the Sway 1.7 and wlroots 0.15.1 releases which starts up fine, so clearly a regression somewhere. Will bisect when I get a chance.

**Comment by tokyovigilante** [comment]
```
[Mon  2:19PM] ryan@firebert ~/Projects/Develop/sway/$ git bisect good
9959e6b8dc9b7fbfcccd516e38bc50681f38e275 is the first bad commit
commit 9959e6b8dc9b7fbfcccd516e38bc50681f38e275
Author: Simon Ser <contact@emersion.fr>
Date:   Wed Feb 1 20:24:40 2023 +0100

    Convert to *_try_from_wlr_surface()

    References: https://gitlab.freedesktop.org/wlroots/wlroots/-/merge_requests/3991

 sway/input/cursor.c         |  9 +++------
 sway/input/seatop_default.c | 18 ++++++++----------
 sway/tree/container.c       | 10 +++++-----
 sway/tree/view.c            | 20 ++++++--------------
 sway/xdg_activation_v1.c    |  6 +-----
 5 files changed, 23 insertions(+), 40 deletions(-)
```

**Comment by bl4ckb0ne** [comment]
What's the behaviour you're experiencing? Do you have a stacktrace?

**Comment by tokyovigilante** [comment]
As discussed above starting sway by running `sway` from a tty just locks up the tty, with the `sway` process running. I can switch to another TTY and attach a debugger, with the log output and backtrace as above. It seems to be waiting in `epoll_wait` possibly as intended in the event loop, but does not switch to graphical mode. The bisect above is the first buggy commit, with the commit before starting up fine (needing a slightly older wlroots).

**Comment by rpigott** [comment]
b31e97e55d76 should build on at least wlroots fbf5982e3838^..711a1a3ed421^

9959e6b8dc9b should build on at least wlroots eb3e8f08a8bb^..0c966f102c1d^

Those ranges don't overlap, what wlroots version did you use for 9959e6b8dc9b (first bad) and b31e97e55d76 (last good)? I'm feeling the regression is more likely within eb3e8f08a8bb^..0c966f102c1d^. In particular, does 9959e6b8dc9b+wlroots@eb3e8f08a8bb still fail?

**Comment by tokyovigilante** [comment]
I have been a bit inobservant. On my laptop,

For b31e97e: wlroots fbf5892
for 9959e6b: wlroots 0534d12 (noting that I arrived at this commit using the tried and tested method of `git checkout HEAD-10` starting at master until the build succeeded)

Testing 9959e6b with additional wlroots commits:
wlroots@eb3e8f0: works!
wlroots@0c966f1: build fails (to new, missing `wlr_linux_dmabuf_feedback_v1` struct members removed in 5cd14dfc)
wlroots@f6008ff: fails!

So it would seem this is actually a wlroots regression between eb3e8f0 and f6008ff, so a quick `git bisect` later:
```
[Tue  7:47PM] ryan@firebert ~/Projects/Develop/sway/subprojects/wlroots/$ git bisect good
8df62e4016ebb633ab281c9dce4583422a070cfa is the first bad commit
commit 8df62e4016ebb633ab281c9dce4583422a070cfa
Author: Simon Ser <contact@emersion.fr>
Date:   Thu Feb 2 11:58:07 2023 +0100

    backend/drm: require hwdata pkg-config file

    The hardcoded fallback "/usr/share/hwdata/pnp.ids" was only a
    temporary solution to get things working while distributions were
    still working on shipping it.

 backend/drm/meson.build | 20 ++++++++++----------
 1 file changed, 10 insertions(+), 10 deletions(-)
```
And of course, my desktop has both `hwdata` and `hwdata-devel` installed, but my laptop only has `hwdata`, missing the .pc file.
Installing the `-devel` package allows wlroots `HEAD` to enable the DRM backend and run.

Overall, not a bug, and definitely not a sway bug, and I should read meson's output more carefully.
```
wlroots| Build-time dependency hwdata found: NO (tried pkgconfig and cmake)
wlroots| Message: Required for the DRM backend.

wlroots 0.17.0-dev

    drm-backend     : NO
```

Closing, but would anyone really want to build without the DRM backend (I suppose running nested in Gnome etc)? If so, I will see myself out. Thanks all for the help and making me do a proper investigation.

**Comment by rpigott** [comment]
Nice! Glad you got it working.

**Comment by emersion** [comment]
I've opened [wlroots!4033](https://gitlab.freedesktop.org/wlroots/wlroots/-/merge_requests/4033) to improve the situation when the DRM backend is disabled at compile-time.
